### PR TITLE
Close #12645: Add sync engine aware debouncing

### LINKED ISSUE #12645: Sync debouncing should respect SyncEngines
Now that `syncNow` supports specifying sync engines, the debounce mechanism in the SyncManager should only debounce requests from the same engine. For example:

- Tabs sync is initiated
- Full sync is initiated with `debounce = true`

The full sync will now be debounced, even if the intention was to only debounce if a full sync had occured within the time out.

Each engine should be separately tracked so that it can be separately debounced, and full syncs should only be debounced if every supported engine has been debounced within the timer.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21284)

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
Fixes #12645

### CODE REVIEW

**Comment by rocketsroger** [comment]
For a cleaner review, I didn't fix the trailing comma issue for ktlint.  I will fix ktlint in a separate PR once this is approved.

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (rocketsroger) [inline components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:530]: Note that this is now 5 seconds, not 5 minutes.  Also I wonder if I should rename it.

Comment (MatthewTighe) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:530]: I'll give this a more thorough review soon, once wallpapers have calmed down a little bit. Please remind me if it takes too long.

For now, I'm curious whether we want the default to be this short. It seems likely that the majority of syncs will come from the home screen, especially given that they will trigger when new tabs are opened or the awesomebar is used. 

I'm not sure I have a strong opinion here, but I'm wondering if it makes sense to keep the default debounce timer a little longer and instead configure the home screen debounce specifically to be shorter assuming we also tackle https://github.com/mozilla-mobile/android-components/issues/12646

Comment (rocketsroger) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:530]: > For now, I'm curious whether we want the default to be this short. It seems likely that the majority of syncs will come from the home screen, especially given that they will trigger when new tabs are opened or the awesomebar is used.

I chose 5 seconds as default value because of the requirement.  From testing I frequently see multiple tab sync calls fairly close to each other (within a second).  This is a good first step so at least we don't sync too closely.

> ... but I'm wondering if it makes sense to keep the default debounce timer a little longer and instead configure the home screen debounce specifically to be shorter assuming we also tackle #12646

Agreed.  But I think we should do one thing at a time.  For this issue we get debounce to work correctly by taking sync engines into account.  After that we can think about a design to give callers ways to specify debounce time in #12646.

**Comment by rocketsroger** [comment]
@MatthewTighe, @jonalmeida Is this something we still want to move forward?  I feel that we should at least first prevent close together syncs before we design/add support for caller to specify custom debounce time.

**Comment by MatthewTighe** [comment]
> Is this something we still want to move forward? I feel that we should at least first prevent close together syncs before we design/add support for caller to specify custom debounce time. 

Thanks for the reminder about this! I think we should, and I'll plan to give this a thorough review either later today or tomorrow. Been trying to wrap up the thing I'm on before context-switching fully in to Sync again, as I'd like to make some plans for task continuity improvements for Q4/Q1.

File: components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt
Comment (MatthewTighe) [inline components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt:75]:
```suggestion
    fun `GIVEN work is set to be debounced THEN last synced timestamp is compared to buffer`() {
```

File: components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt
Comment (MatthewTighe) [inline components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt:89]:
```suggestion
    fun `GIVEN work is set to be debounced WHEN there is not a saved time stamp THEN work will not be debounced`() {
```

File: components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt
Comment (MatthewTighe) [inline components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt:70]: Little bit of a rant incoming. No need for changes, just some thoughts that might be worth sharing.

Generally speaking, I prefer to focus on tests that cover the smallest possible public API. This creates expectations about the behavior of the class as it would actually be used. Marking functions as internal in order to test them makes a little more sense in a modular architecture like we have in A-C, as the `internal` keyword is still respected across module boundaries. I find it to be a bit of an anti-pattern in Fenix, though, where the entire project is still a single module and all those `internal` functions become part of the global namespace. 

That said, the existing structure of this code is really hard to test well. That's always a bit of a problem when using Android platform types like `WorkManager`, but I think we could make pretty huge improvements in that regard if we get the chance to refactor more of this in the future. So any increase in test coverage here is a good thing in my books 🙂

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (jonalmeida) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:533]: I think the in-memory fix might work for now to avoid re-syncing while actively using the application. `companion object` is a thread-safe singleton so it'll live for the lifetime of the app. If the app is killed and we lose the in-memory mapping, then we've not used the application and synced a lot to worry about debouncing.

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (jonalmeida) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:300]: Yak-shaving: don't hate on the early-returns as an option too.

```kotlin
if (isDebounched()) {
  return false
}

engineSyncTimestamp[engine]?.let {
    (System.currentTimeMillis() - it) < SYNC_STAGGER_BUFFER_MS
} ?: false
```

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (jonalmeida) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:533]: If you do decide to persist the last synced values, please use the `SharedPreferencesCache` for serializing and deserializing.

Comment (rocketsroger) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:533]: That was my thought as well.  Since closing the app and reopening will usually take some time, persisting the value is not needed yet.  When we add the API to allow user to customize custom debounce time, we should discuss if/how we persist them.

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (rocketsroger) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:300]: Agreed. I'll update.  Thanks

File: components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt
Comment (rocketsroger) [reply components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt:61]: Good catch, I'll update.  Thanks

File: components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt
Comment (rocketsroger) [reply components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt:89]: I'll update.  Thanks

File: components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt
Comment (rocketsroger) [reply components/service/firefox-accounts/src/test/java/mozilla/components/service/fxa/sync/WorkManagerSyncManagerTest.kt:70]: Agreed.  The first time I did similar thing for testing I was uncomfortable as well.  I wish tests by default can call into `private` functions.

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (rocketsroger) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:325]: Agreed, especially once we allow users to customize debounce time.  I will take this into account when we support custom debounce time.

Comment (rocketsroger) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:325]: I've added these notes to the follow up issue.  Thanks

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (MatthewTighe) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:530]: > `companion object` is a thread-safe singleton lives for the lifetime of the app

I'm not willing to die on this hill, since this definitely seems the most straightforward way to get the desired behavior. IIUC, however, `static` (and therefore `companion`) objects and properties are not guaranteed to live in memory for the lifetime of an application. 

`companion` objects are generated as `static` members of their parent class, and will only remain in memory while the class definition is loaded. It is apparently unlikely for classes to ever be unloaded on a standard JVM, but I don't understand the ART runtime well enough to say the same. I would worry that Android's aggressive memory management is more likely to unload disused classes.

I would guess that the likelihood of that actually happening is incredibly low, if even possible. This seems especially true since I would guess that the regularity of syncs would keep these classes prioritized and loaded even _if_ any of my assumptions about ART are true

Comment (MatthewTighe) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:530]: I'm also curious about the `thread-safe singleton`. I think they are thread-safe during initialization, but I don't think the properties defined on them are. Since `doWork` is suspending on a `Dispatcher` with multiple threads, is it possible we could hit a deadlock without making this property `synchronized`?

Comment (rocketsroger) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:530]: Thread safety is concerning.  I'll look into this.  Thanks

**Comment by rocketsroger** [comment]
I've removed the ktlint fixes with the assumption that this change will wait for the project wide ktlint fixes to land first.

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt
Comment (MatthewTighe) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt]: Oh good, glad there was an easy fix for that.

Also, just for the sake of the conversation: my underlying issue here isn't about having persisted state of the timestamps. I just think it's a little code-smelly to be using a shared static property as a cache. Writing the state to disk felt like the easy solution to put the cache in a shared layer, which in this case would be disk storage. Something like an injected in-memory cache whose lifetime was controlled further up the object hierarchy could also work:

```
internal class WorkManagerSyncWorker(
    private val context: Context,
    private val params: WorkerParameters,
    private val syncTimestampCache: InMemoryCache,
) : CoroutineWorker(context, params) {

...
// the lifetime of this object is known to be longer, and can share the cache between periodic and one-time sync workers
internal class WorkManagerSyncDispatcher {
    private val timestampCache: InMemoryCache = mutableMapOf(<engines>, <timestamps>)
}
```

**Review by MatthewTighe — APPROVED** [review]
LGTM given the changes specified in https://github.com/mozilla-mobile/android-components/issues/12646. Given how small the debounce window has become in this, I don't think we're really at risk to have sync failures block future successes as mentioned in [this comment](https://github.com/mozilla-mobile/android-components/pull/12790/files#r993952935)

I do think we should prioritize configurable debounce times at the call-site, so that we can re-increase the default. I'm a little concerned the current value could generate an unexpected amount of traffic somehow. I would be happy to work on that if you're otherwise busy, just LMK

**Comment by rocketsroger** [comment]
> I do think we should prioritize configurable debounce times at the call-site, so that we can re-increase the default. I'm a little concerned the current value could generate an unexpected amount of traffic somehow. I would be happy to work on that if you're otherwise busy, just LMK

Sure, please let me know when you want to start looking into it and maybe we can look at the design together.  Thanks

**Pull request merged**
